# Smart Traffic Management System

## 01 — Dataset Understanding

### Objective

The objective of this notebook is to understand the structure,
schema, data types, and basic characteristics of the Smart
Traffic Management System dataset before performing detailed
Exploratory Data Analysis (EDA).

### This notebook will focus on

- Loading a manageable sample of the large dataset
- Understanding the dataset structure
- Inspecting rows and columns
- Understanding data types
- Measuring memory usage
- Identifying missing values
- Identifying duplicate rows
- Understanding unique values
- Classifying columns
- Creating an initial data dictionary
- Recording initial observations

### Important

The complete dataset is approximately 9.74 GB and therefore
will NOT be loaded into memory at this stage.

A representative sample of rows will be used for initial
dataset understanding.

The full dataset will later be processed using chunk-based
processing where required.

## 2. Import Required Libraries

We will use:

- pandas → Data manipulation and analysis
- numpy → Numerical operations
- pathlib → Cross-platform file path handling
- matplotlib → Visualization support for later analysis
- seaborn → Statistical visualization support for later analysis

In [54]:
import pandas as pd
import numpy as np

from pathlib import Path

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

## Load Dataset

In [55]:
DATA_PATH = Path("../data/processed/traffic_2025.csv")

if not DATA_PATH.exists():
    raise FileNotFoundError(
        f"Dataset not found: {DATA_PATH.resolve()}"
    )

df = pd.read_csv(
    DATA_PATH,
    nrows=10_000
)

print(f"Shape: {df.shape}")

Shape: (10000, 103)


In [42]:
df.head()

,NB_SCATS_SITE,QT_INTERVAL_COUNT,NB_DETECTOR,V00,V01,V02,V03,V04,V05,V06,V07,V08,V09,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,V29,V30,V31,V32,V33,V34,V35,V36,V37,V38,V39,V40,V41,V42,V43,V44,V45,V46,V47,V48,V49,V50,V51,V52,V53,V54,V55,V56,V57,V58,V59,V60,V61,V62,V63,V64,V65,V66,V67,V68,V69,V70,V71,V72,V73,V74,V75,V76,V77,V78,V79,V80,V81,V82,V83,V84,V85,V86,V87,V88,V89,V90,V91,V92,V93,V94,V95,NM_REGION,CT_RECORDS,QT_VOLUME_24HOUR,CT_ALARM_24HOUR
0,100,2025-04-01,1,2,1,1,2,4,1,4,0,1,1,1,0,1,0,0,1,0,0,1,0,3,4,4,3,7,12,16,11,5,9,15,12,26,13,20,20,21,14,9,11,10,7,14,16,14,14,11,17,17,22,17,15,14,19,11,12,29,20,26,11,15,13,14,18,17,15,19,16,21,12,15,14,19,17,15,10,11,7,13,8,12,5,3,12,4,10,8,18,11,10,12,9,4,7,3,3,WV1,96,992,0
1,100,2025-04-01,2,12,18,4,9,5,5,7,3,3,5,1,0,4,0,1,2,1,1,2,5,7,15,32,22,31,50,68,68,62,84,102,130,121,143,167,165,141,113,115,108,116,107,102,119,111,104,140,135,130,115,119,115,127,128,113,112,159,139,175,153,168,201,200,195,207,232,219,234,234,230,237,220,203,183,185,149,113,115,100,100,75,99,69,86,82,66,46,63,43,44,28,32,29,22,13,8,WV1,96,8881,0
2,100,2025-04-01,3,22,8,8,10,8,10,8,5,9,4,6,4,8,10,10,6,12,7,13,15,14,23,41,43,53,56,87,81,70,99,95,103,145,168,158,172,133,108,110,115,108,110,103,97,94,83,103,121,132,114,115,121,114,117,119,117,136,153,149,156,172,188,189,192,203,201,215,241,220,209,204,202,193,178,135,139,116,110,112,82,80,95,66,87,83,68,69,72,64,62,58,55,48,29,21,28,WV1,96,8905,0
3,100,2025-04-01,4,17,9,9,6,5,6,3,4,2,3,3,1,2,2,4,3,0,5,6,5,8,13,30,29,25,31,73,64,77,82,115,125,128,154,179,169,123,125,107,106,103,113,94,100,108,102,109,101,128,110,111,109,114,102,113,104,162,146,155,166,190,189,210,210,190,207,208,236,220,222,219,200,190,181,140,134,113,108,81,74,84,97,71,83,79,71,45,53,44,48,38,34,22,18,13,17,WV1,96,8547,0
4,100,2025-04-01,5,3,3,2,1,2,4,0,0,1,2,0,1,0,1,0,1,0,0,1,3,4,4,3,9,7,13,17,16,12,6,9,13,16,17,25,21,20,13,8,11,7,12,7,13,13,18,17,18,18,8,13,11,6,13,16,17,18,5,14,13,24,25,20,14,21,12,12,16,20,28,16,15,16,15,10,20,11,14,7,21,9,20,14,12,7,9,11,9,6,4,5,3,14,3,3,2,WV1,96,994,0


In [43]:
df.tail()

,NB_SCATS_SITE,QT_INTERVAL_COUNT,NB_DETECTOR,V00,V01,V02,V03,V04,V05,V06,V07,V08,V09,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,V29,V30,V31,V32,V33,V34,V35,V36,V37,V38,V39,V40,V41,V42,V43,V44,V45,V46,V47,V48,V49,V50,V51,V52,V53,V54,V55,V56,V57,V58,V59,V60,V61,V62,V63,V64,V65,V66,V67,V68,V69,V70,V71,V72,V73,V74,V75,V76,V77,V78,V79,V80,V81,V82,V83,V84,V85,V86,V87,V88,V89,V90,V91,V92,V93,V94,V95,NM_REGION,CT_RECORDS,QT_VOLUME_24HOUR,CT_ALARM_24HOUR
9995,548,2025-04-01,12,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,FRA,96,0,0
9996,548,2025-04-01,13,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,FRA,96,0,0
9997,548,2025-04-01,14,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,FRA,96,0,0
9998,548,2025-04-01,15,1,2,0,1,1,0,1,2,0,0,1,0,0,0,0,0,0,0,1,0,0,1,0,0,1,0,0,0,0,3,4,14,8,9,2,11,24,9,9,16,9,16,18,17,16,15,20,22,14,11,24,20,20,22,26,19,11,15,22,11,13,15,20,14,17,10,16,14,18,21,22,27,14,21,15,12,6,17,8,9,12,10,9,8,6,5,7,3,2,4,2,3,5,3,1,1,FRA,96,859,0
9999,548,2025-04-01,16,6,7,6,3,2,0,1,4,1,1,2,0,0,3,2,8,7,12,24,36,41,72,117,123,125,131,127,130,145,145,158,169,164,145,128,162,113,102,98,97,85,76,87,83,72,86,84,78,81,94,77,78,83,89,81,81,87,93,95,82,78,107,97,105,79,106,105,115,102,85,114,75,65,72,73,69,56,60,60,57,53,47,38,36,30,45,37,33,23,34,13,21,18,13,11,5,FRA,96,6426,0


In [44]:
df.sample(5)

,NB_SCATS_SITE,QT_INTERVAL_COUNT,NB_DETECTOR,V00,V01,V02,V03,V04,V05,V06,V07,V08,V09,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,V29,V30,V31,V32,V33,V34,V35,V36,V37,V38,V39,V40,V41,V42,V43,V44,V45,V46,V47,V48,V49,V50,V51,V52,V53,V54,V55,V56,V57,V58,V59,V60,V61,V62,V63,V64,V65,V66,V67,V68,V69,V70,V71,V72,V73,V74,V75,V76,V77,V78,V79,V80,V81,V82,V83,V84,V85,V86,V87,V88,V89,V90,V91,V92,V93,V94,V95,NM_REGION,CT_RECORDS,QT_VOLUME_24HOUR,CT_ALARM_24HOUR
2412,205,2025-04-01,21,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,SPR,96,0,0
4264,285,2025-04-01,17,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,SPR,96,0,0
5903,362,2025-04-01,15,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,WV1,96,0,0
3261,243,2025-04-01,14,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,FR2,96,0,0
6613,397,2025-04-01,14,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,FRA,96,0,0


## Basic Inspection


In [59]:
print("Shape:", df.shape)





Shape: (10000, 103)


In [60]:
print(df.columns.tolist())

['NB_SCATS_SITE', 'QT_INTERVAL_COUNT', 'NB_DETECTOR', 'V00', 'V01', 'V02', 'V03', 'V04', 'V05', 'V06', 'V07', 'V08', 'V09', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31', 'V32', 'V33', 'V34', 'V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41', 'V42', 'V43', 'V44', 'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51', 'V52', 'V53', 'V54', 'V55', 'V56', 'V57', 'V58', 'V59', 'V60', 'V61', 'V62', 'V63', 'V64', 'V65', 'V66', 'V67', 'V68', 'V69', 'V70', 'V71', 'V72', 'V73', 'V74', 'V75', 'V76', 'V77', 'V78', 'V79', 'V80', 'V81', 'V82', 'V83', 'V84', 'V85', 'V86', 'V87', 'V88', 'V89', 'V90', 'V91', 'V92', 'V93', 'V94', 'V95', 'NM_REGION', 'CT_RECORDS', 'QT_VOLUME_24HOUR', 'CT_ALARM_24HOUR']


In [61]:
df.dtypes

NB_SCATS_SITE        int64
QT_INTERVAL_COUNT      str
NB_DETECTOR          int64
V00                  int64
V01                  int64
                     ...  
V95                  int64
NM_REGION              str
CT_RECORDS           int64
QT_VOLUME_24HOUR     int64
CT_ALARM_24HOUR      int64
Length: 103, dtype: object

In [48]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Columns: 103 entries, NB_SCATS_SITE to CT_ALARM_24HOUR
dtypes: int64(101), str(2)
memory usage: 7.9 MB


In [49]:
df.info(verbose=True)

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 103 columns):
 #    Column             Dtype
---   ------             -----
 0    NB_SCATS_SITE      int64
 1    QT_INTERVAL_COUNT  str  
 2    NB_DETECTOR        int64
 3    V00                int64
 4    V01                int64
 5    V02                int64
 6    V03                int64
 7    V04                int64
 8    V05                int64
 9    V06                int64
 10   V07                int64
 11   V08                int64
 12   V09                int64
 13   V10                int64
 14   V11                int64
 15   V12                int64
 16   V13                int64
 17   V14                int64
 18   V15                int64
 19   V16                int64
 20   V17                int64
 21   V18                int64
 22   V19                int64
 23   V20                int64
 24   V21                int64
 25   V22                int64
 26   V23                int64
 27

In [50]:
df.describe()

,NB_SCATS_SITE,NB_DETECTOR,V00,V01,V02,V03,V04,V05,V06,V07,V08,V09,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,V29,V30,V31,V32,V33,V34,V35,V36,V37,V38,V39,V40,V41,V42,V43,V44,V45,V46,V47,V48,V49,V50,V51,V52,V53,V54,V55,V56,V57,V58,V59,V60,V61,V62,V63,V64,V65,V66,V67,V68,V69,V70,V71,V72,V73,V74,V75,V76,V77,V78,V79,V80,V81,V82,V83,V84,V85,V86,V87,V88,V89,V90,V91,V92,V93,V94,V95,CT_RECORDS,QT_VOLUME_24HOUR,CT_ALARM_24HOUR
count,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.00000,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.0,10000.00000,10000.000000
mean,322.427200,13.098400,3.119200,2.729300,2.457800,2.202400,1.957700,1.872000,1.801400,1.713200,1.738800,1.66170,1.559700,1.536100,1.546700,1.615500,1.800600,2.012100,2.128400,2.458300,3.358300,4.022800,5.049500,7.381900,10.40110,11.970600,13.451700,16.847700,19.895100,21.198100,21.479700,24.792200,27.315900,29.767800,30.217300,31.863600,32.305600,31.549400,28.184100,25.332000,23.254800,22.822700,21.443800,21.801100,21.727000,22.043900,21.792500,22.22230,22.526900,22.955500,23.069000,23.484200,23.533500,23.376400,23.282300,23.69440,23.473800,23.765100,24.459300,26.175800,27.490500,28.61820,30.294100,31.31310,31.886400,32.131900,31.886100,31.992900,32.210300,32.296300,33.11440,33.06170,31.66420,30.036600,27.636100,25.682000,22.895400,21.250300,18.839100,17.407000,16.117700,15.323400,14.341200,13.522200,12.741600,12.284100,11.702100,10.739200,9.686300,9.195100,8.910000,7.951300,6.829200,5.701800,5.123900,4.316000,3.617000,2.963100,96.0,1615.49610,1.104000
std,129.808413,7.542414,19.613563,19.672964,20.416648,19.130971,19.476264,20.813744,20.334417,21.239111,22.604413,22.49167,21.196791,21.642947,21.286525,21.870174,22.445539,23.363992,22.178681,21.471143,20.902641,19.337541,20.176803,25.047391,29.69438,30.282051,32.095259,37.501186,42.014024,42.970659,44.631058,49.739803,52.515296,54.789531,54.166865,56.064446,55.461206,54.217205,48.840532,44.681562,40.912268,39.776844,36.956131,37.772906,38.034289,39.300522,39.313077,39.23136,39.834726,40.425014,40.780921,41.150274,41.747879,41.471724,41.270439,42.26628,41.201362,41.671885,42.930528,46.329288,48.136093,49.80684,51.488087,53.43199,53.900353,54.842677,54.799601,54.754933,55.099509,55.239468,56.37083,56.35641,54.40141,52.832528,49.158219,45.629798,41.006485,38.804804,34.825817,32.054867,29.325887,27.906245,26.521375,25.422737,24.282945,23.967537,23.115505,21.587857,20.088416,19.603369,19.270413,17.943899,16.600018,14.835474,14.019351,12.399118,12.644426,11.618237,0.0,2834.12222,10.002009
min,100.000000,0.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.00000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.000000,-1.00000,-1.000000,-1.000000

## Data Quality Check

In [63]:
missing = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_percent": df.isnull().mean() * 100
})

missing = missing.sort_values(
    "missing_count",
    ascending=False
)

missing[missing["missing_count"] > 0]

,missing_count,missing_percent


In [64]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [65]:
unique_counts = (
    df.nunique(dropna=False)
      .sort_values()
)

unique_counts

QT_INTERVAL_COUNT       1
CT_RECORDS              1
NM_REGION              14
CT_ALARM_24HOUR        19
NB_DETECTOR            32
                     ... 
V66                   292
V69                   293
V31                   294
NB_SCATS_SITE         396
QT_VOLUME_24HOUR     3413
Length: 103, dtype: int64

In [66]:
print("Unique SCATS sites:",
      df["NB_SCATS_SITE"].nunique())

print("Unique detectors:",
      df["NB_DETECTOR"].nunique())

print("Unique regions:",
      df["NM_REGION"].nunique())

Unique SCATS sites: 396
Unique detectors: 32
Unique regions: 14


## Traffic Volume Columns

In [67]:
traffic_columns = [
    col for col in df.columns
    if col.startswith("V") and col[1:].isdigit()
]

traffic_columns = sorted(
    traffic_columns,
    key=lambda x: int(x[1:])
)

print("Number of traffic columns:", len(traffic_columns))
print(traffic_columns)

Number of traffic columns: 96
['V00', 'V01', 'V02', 'V03', 'V04', 'V05', 'V06', 'V07', 'V08', 'V09', 'V10', 'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31', 'V32', 'V33', 'V34', 'V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41', 'V42', 'V43', 'V44', 'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51', 'V52', 'V53', 'V54', 'V55', 'V56', 'V57', 'V58', 'V59', 'V60', 'V61', 'V62', 'V63', 'V64', 'V65', 'V66', 'V67', 'V68', 'V69', 'V70', 'V71', 'V72', 'V73', 'V74', 'V75', 'V76', 'V77', 'V78', 'V79', 'V80', 'V81', 'V82', 'V83', 'V84', 'V85', 'V86', 'V87', 'V88', 'V89', 'V90', 'V91', 'V92', 'V93', 'V94', 'V95']


In [68]:
df[traffic_columns].head()

,V00,V01,V02,V03,V04,V05,V06,V07,V08,V09,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,V29,V30,V31,V32,V33,V34,V35,V36,V37,V38,V39,V40,V41,V42,V43,V44,V45,V46,V47,V48,V49,V50,V51,V52,V53,V54,V55,V56,V57,V58,V59,V60,V61,V62,V63,V64,V65,V66,V67,V68,V69,V70,V71,V72,V73,V74,V75,V76,V77,V78,V79,V80,V81,V82,V83,V84,V85,V86,V87,V88,V89,V90,V91,V92,V93,V94,V95
0,2,1,1,2,4,1,4,0,1,1,1,0,1,0,0,1,0,0,1,0,3,4,4,3,7,12,16,11,5,9,15,12,26,13,20,20,21,14,9,11,10,7,14,16,14,14,11,17,17,22,17,15,14,19,11,12,29,20,26,11,15,13,14,18,17,15,19,16,21,12,15,14,19,17,15,10,11,7,13,8,12,5,3,12,4,10,8,18,11,10,12,9,4,7,3,3
1,12,18,4,9,5,5,7,3,3,5,1,0,4,0,1,2,1,1,2,5,7,15,32,22,31,50,68,68,62,84,102,130,121,143,167,165,141,113,115,108,116,107,102,119,111,104,140,135,130,115,119,115,127,128,113,112,159,139,175,153,168,201,200,195,207,232,219,234,234,230,237,220,203,183,185,149,113,115,100,100,75,99,69,86,82,66,46,63,43,44,28,32,29,22,13,8
2,22,8,8,10,8,10,8,5,9,4,6,4,8,10,10,6,12,7,13,15,14,23,41,43,53,56,87,81,70,99,95,103,145,168,158,172,133,108,110,115,108,110,103,97,94,83,103,121,132,114,115,121,114,117,119,117,136,153,149,156,172,188,189,192,203,201,215,241,220,209,204,202,193,178,135,139,116,110,112,82,80,95,66,87,83,68,69,72,64,62,58,55,48,29,21,28
3,17,9,9,6,5,6,3,4,2,3,3,1,2,2,4,3,0,5,6,5,8,13,30,29,25,31,73,64,77,82,115,125,128,154,179,169,123,125,107,106,103,113,94,100,108,102,109,101,128,110,111,109,114,102,113,104,162,146,155,166,190,189,210,210,190,207,208,236,220,222,219,200,190,181,140,134,113,108,81,74,84,97,71,83,79,71,45,53,44,48,38,34,22,18,13,17
4,3,3,2,1,2,4,0,0,1,2,0,1,0,1,0,1,0,0,1,3,4,4,3,9,7,13,17,16,12,6,9,13,16,17,25,21,20,13,8,11,7,12,7,13,13,18,17,18,18,8,13,11,6,13,16,17,18,5,14,13,24,25,20,14,21,12,12,16,20,28,16,15,16,15,10,20,11,14,7,21,9,20,14,12,7,9,11,9,6,4,5,3,14,3,3,2


In [69]:
df[traffic_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
V00,10000.0,3.1192,19.613563,-1.0,0.0,0.0,2.00,759.0
V01,10000.0,2.7293,19.672964,-1.0,0.0,0.0,2.00,759.0
V02,10000.0,2.4578,20.416648,-1.0,0.0,0.0,1.00,759.0
V03,10000.0,2.2024,19.130971,-1.0,0.0,0.0,1.00,759.0
V04,10000.0,1.9577,19.476264,-1.0,0.0,0.0,1.00,759.0
V05,10000.0,1.8720,20.813744,-1.0,0.0,0.0,1.00,759.0
V06,10000.0,1.8014,20.334417,-1.0,0.0,0.0,1.00,759.0
V07,10000.0,1.7132,21.239111,-1.0,0.0,0.0,1.00,759.0
V08,10000.0,1.7388,22.604413,-1.0,0.0,0.0,0.00,759.0
V09,10000.0,1.6617,22.491670,-1.0,0.0,0.0,0.00,759.0


In [70]:
df[traffic_columns].dtypes.value_counts()

int64    96
Name: count, dtype: int64

In [71]:
negative_values = (df[traffic_columns] < 0).sum()

negative_values[negative_values > 0]

V00    112
V01    112
V02    112
V03    112
V04    112
V05    112
V06    112
V07    112
V08    112
V09    112
V10    112
V11    112
V12    112
V13    112
V14    112
V15    112
V16    112
V17    112
V18    112
V19    112
V20    136
V21    136
V22    112
V23    112
V24    112
V25    112
V26    136
V27    136
V28    112
V29    137
V30    136
V31    112
V32    112
V33    112
V34    112
V35    120
V36    119
V37    115
V38    115
V39    115
V40    115
V41    114
V42    111
V43    114
V44    114
V45    113
V46    111
V47    116
V48    116
V49    114
V50    112
V51    116
V52    115
V53    115
V54    114
V55    116
V56    115
V57    115
V58    114
V59    114
V60    114
V61    113
V62    113
V63    114
V64    113
V65    113
V66    113
V67    118
V68    118
V69    118
V70    116
V71    131
V72    115
V73    115
V74    114
V75    115
V76    114
V77    113
V78    112
V79    112
V80    111
V81    111
V82    111
V83    111
V84    111
V85    111
V86    135
V87    114
V88    111
V89    111
V90    111

In [72]:
zero_values = (df[traffic_columns] == 0).sum()

zero_values.sort_values(ascending=False)

V11    7544
V12    7443
V10    7434
V09    7431
V08    7393
V13    7384
V14    7323
V07    7222
V15    7212
V06    7145
V05    7120
V16    7032
V04    7006
V03    6903
V17    6882
V02    6794
V01    6664
V18    6660
V00    6479
V95    6453
V19    6419
V94    6328
V20    6281
V93    6218
V21    6096
V92    6075
V91    6012
V90    5964
V22    5947
V89    5879
V23    5851
V88    5802
V24    5792
V87    5779
V25    5719
V86    5717
V85    5694
V26    5655
V84    5654
V83    5633
V82    5614
V81    5602
V27    5601
V28    5582
V79    5566
V80    5559
V29    5543
V78    5543
V30    5540
V76    5525
V31    5516
V77    5513
V73    5490
V43    5488
V75    5483
V32    5479
V44    5478
V45    5478
V46    5476
V50    5476
V47    5473
V74    5473
V34    5473
V42    5472
V56    5471
V48    5469
V72    5469
V71    5469
V54    5469
V53    5468
V41    5467
V33    5466
V39    5464
V38    5463
V57    5461
V37    5461
V70    5459
V36    5458
V64    5458
V55    5458
V40    5457
V52    5456
V59    5455
V49 

## Important Columns

In [73]:
df[
    [
        "NB_SCATS_SITE",
        "QT_INTERVAL_COUNT",
        "NB_DETECTOR",
        "NM_REGION",
        "CT_RECORDS",
        "QT_VOLUME_24HOUR",
        "CT_ALARM_24HOUR"
    ]
].head(10)

,NB_SCATS_SITE,QT_INTERVAL_COUNT,NB_DETECTOR,NM_REGION,CT_RECORDS,QT_VOLUME_24HOUR,CT_ALARM_24HOUR
0,100,2025-04-01,1,WV1,96,992,0
1,100,2025-04-01,2,WV1,96,8881,0
2,100,2025-04-01,3,WV1,96,8905,0
3,100,2025-04-01,4,WV1,96,8547,0
4,100,2025-04-01,5,WV1,96,994,0
5,100,2025-04-01,6,WV1,96,0,0
6,100,2025-04-01,7,WV1,96,0,0
7,100,2025-04-01,8,WV1,96,0,0
8,100,2025-04-01,9,WV1,96,0,0
9,100,2025-04-01,10,WV1,96,0,0


In [74]:
df["QT_INTERVAL_COUNT"].value_counts().sort_index()

QT_INTERVAL_COUNT
2025-04-01    10000
Name: count, dtype: int64

In [75]:
df["NM_REGION"].value_counts(dropna=False)

NM_REGION
WV1    1848
FRA    1784
CRN    1240
WV2    1144
SPR    1144
BBN     928
DON     616
SP2     432
MEN     416
FR2     328
VIC      48
DUP      24
DIO      24
VI2      24
Name: count, dtype: int64

In [76]:
interval_sum = df[traffic_columns].sum(axis=1)

comparison = pd.DataFrame({
    "interval_sum": interval_sum,
    "QT_VOLUME_24HOUR": df["QT_VOLUME_24HOUR"]
})

comparison["difference"] = (
    comparison["interval_sum"]
    - comparison["QT_VOLUME_24HOUR"]
)

comparison.head()

,interval_sum,QT_VOLUME_24HOUR,difference
0,992,992,0
1,8881,8881,0
2,8905,8905,0
3,8547,8547,0
4,994,994,0


In [77]:
print(
    "Rows with different values:",
    (comparison["difference"] != 0).sum()
)

print(
    "Maximum absolute difference:",
    comparison["difference"].abs().max()
)

Rows with different values: 202
Maximum absolute difference: 9955


## Initial Observations

- The dataset contains 103 columns.
- V00–V95 form a group of 96 traffic-related columns.
- NB_SCATS_SITE and NB_DETECTOR appear to identify traffic infrastructure.
- NM_REGION contains regional information.
- QT_VOLUME_24HOUR is an aggregate traffic-related field.
- Missing values and duplicate records need to be investigated further.
- The relationship between V00–V95 and QT_VOLUME_24HOUR was checked rather than assumed.